### ISSUE TITLE
Add optional JAX support for differentiable numerical kernels

### ISSUE THREAD

**Issue by HongSik-Yun-Fusion** [VEST-Tokamak/vaft#1002]
## Summary

Prepare **optional JAX support in VAFT** for a small set of differentiable numerical kernels, without making JAX part of the default dependency stack and without turning VAFT into a JAX-first framework.

The intended model is:

```text
normal VAFT
Python 3.10–3.13
NumPy / SciPy
        │
        └── canonical production implementation

optional JAX environment
Python 3.12
vaft[jax]
        │
        └── selected JAX-compatible numerical kernels
            for jit / vmap / autodiff use
```

This is primarily infrastructure for downstream solvers such as VFIT, where differentiable inverse problems may benefit from exact Jacobians, batching, JIT compilation, sensitivity analysis, and future joint inference.

The main design requirement is:

> JAX must remain fully opt-in and must not affect ordinary VAFT users.

---

## Motivation

VAFT currently targets a broad Python range and uses NumPy/SciPy as the canonical scientific stack.

That should remain unchanged.

However, some VAFT numerical kernels are natural candidates for a differentiable implementation, especially those used by inverse solvers:

```text
Green functions
    -> point-source electromagnetic response
    -> response matrices
    -> downstream nonlinear inverse problems
```

VFIT is one immediate consumer. Its FourFilament reconstruction repeatedly evaluates magnetic responses as filament positions move during nonlinear least-squares fitting. A JAX-compatible VAFT kernel could allow downstream code to compute exact Jacobians with `jax.jacfwd`, batch many evaluations with `vmap`, and JIT-compile the steady-state numerical path.

This does **not** justify replacing NumPy/SciPy across VAFT.

The purpose of this issue is to establish a narrow optional compatibility layer before downstream projects independently duplicate VAFT physics.

---

## Packaging contract

Keep the existing default VAFT dependency set unchanged.

Add an optional dependency group such as:

```toml
[project.optional-dependencies]
jax = [
    "jax>=0.11,<0.12",
]
```

The exact version range should be chosen against the supported release available at implementation time.

The JAX environment should explicitly target **Python 3.12**.

Example development setup:

```bash
uv venv --python 3.12 .venv-jax
source .venv-jax/bin/activate
uv pip install -e ".[jax]"
```

Ordinary users should continue to install:

```bash
pip install vaft
```

with no JAX dependency, import cost, environment restriction, or behavioral change.

Do not silently skip JAX installation on unsupported Python versions through permissive markers if doing so would make `vaft[jax]` appear successful without actually providing the requested backend. Prefer an explicit environment contract.

---

## Scope

The first implementation should be intentionally small.

### Phase 1 — packaging and compatibility contract

- [ ] Add a `jax` optional dependency group.
- [ ] Document Python 3.12 as the supported JAX environment.
- [ ] Add an actionable import error when a JAX-only API is called without the extra installed.
- [ ] Ensure importing ordinary VAFT never imports JAX.
- [ ] Ensure default tests remain independent of JAX.
- [ ] Define the numerical precision policy for JAX kernels.

### Phase 2 — first physics PoC

Use the exact toroidal-ring electromagnetic Green functions as the initial test case.

Candidate kernels:

```text
green_psi_exact
green_br_exact
green_bz_exact
```

and, if the PoC succeeds:

```text
compute_point_response_matrices
```

The NumPy/SciPy implementation remains the canonical reference.

The JAX implementation must reproduce the same physics conventions, units, signs, singularity handling, and numerical limits.

### Phase 3 — differentiability validation

For the first supported kernels, test:

- `jax.jit`;
- `jax.vmap`;
- `jax.jacfwd`;
- optionally `jax.jacrev` where useful;
- gradients with respect to source coordinates;
- gradients with respect to evaluation coordinates where physically meaningful.

Validate gradients independently against:

- central finite differences;
- analytic derivatives if available;
- NumPy/SciPy reference values.

### Phase 4 — downstream validation

Use VFIT FourFilament as the first downstream scientific benchmark.

The downstream benchmark should compare:

```text
current SciPy solver
vs
SciPy + analytic Jacobian
vs
JAX-compatible VAFT kernel + autodiff Jacobian
```

The JAX layer should be expanded beyond the initial kernels only if the downstream benchmark demonstrates a clear scientific or computational benefit.

---

## Green-function challenge

The exact toroidal-ring Green functions depend on complete elliptic integrals.

This is likely the first non-trivial compatibility issue.

Do not implement the JAX path as a naive mechanical replacement of

```python
import numpy as np
```

with

```python
import jax.numpy as jnp
```

without verifying the required special functions and their differentiability.

The implementation should evaluate the available options explicitly:

1. use a maintained JAX-native elliptic-integral implementation;
2. implement the required complete elliptic integrals in a small validated internal kernel;
3. use custom JVP/VJP rules around a verified implementation;
4. expose analytic derivatives from the canonical physics implementation where that is more robust.

Avoid relying on host callbacks to SciPy as the long-term implementation if they defeat JIT portability or automatic differentiation.

---

## Architecture

Do not introduce a global backend switch throughout VAFT.

Avoid patterns such as:

```python
vaft.set_backend("jax")
```

or pervasive

```python
backend="jax"
```

arguments on unrelated APIs.

Prefer a narrow explicit numerical layer.

A desirable conceptual structure is:

```text
canonical mathematical definition
            │
       shared algebra
        /         \
       /           \
NumPy/SciPy       JAX
production        optional
reference         differentiable
```

The exact package/module layout should be chosen only after the first PoC shows what can actually be shared without making the code harder to audit.

The main architectural rules are:

- NumPy/SciPy remains authoritative;
- JAX support is explicit rather than ambient;
- JAX-specific code stays out of OMAS/IMAS, plotting, database, workflow, and external-code layers;
- no global JAX state is modified at ordinary VAFT import time;
- downstream packages should consume VAFT JAX-compatible physics rather than maintain drifting copies.

---

## Precision policy

Scientific parity must be evaluated in 64-bit floating-point mode.

JAX may default to 32-bit operation depending on configuration, which is not acceptable as an implicit behavior for equilibrium and magnetic-response validation.

Do **not** call a global configuration mutation such as

```python
jax.config.update("jax_enable_x64", True)
```

from ordinary VAFT import code.

Prefer:

- documentation that the JAX environment must enable x64;
- CI with `JAX_ENABLE_X64=1`;
- runtime checks in JAX-only entry points where appropriate;
- tests that explicitly verify expected dtypes and parity.

---

## CI strategy

Do not add JAX to every existing VAFT test job.

Keep the normal CI matrix unchanged.

Add a separate optional job, for example:

```text
JAX compatibility
Python 3.12
CPU-only
JAX_ENABLE_X64=1
    ├── selected JAX unit tests
    ├── NumPy/JAX parity tests
    ├── gradient checks
    └── small performance smoke benchmark
```

GPU CI is out of scope for the first implementation.

Accelerator-specific JAX installation should remain the user's responsibility until VAFT has a demonstrated need to test CUDA/ROCm execution directly.

---

## Candidate initial API surface

The issue should not prescribe final names prematurely, but a successful first stage should provide an explicit way to access JAX-compatible kernels without changing normal VAFT behavior.

Possible shapes include:

```text
vaft.formula.jax.*
```

or a small internal/public submodule dedicated to differentiable kernels.

Do not add JAX branches to every existing function merely for API symmetry.

The canonical NumPy-facing public APIs should continue to work exactly as they do now.

---

## Relationship to analytic derivatives

JAX support should not prevent VAFT from exposing useful analytic derivatives.

For some kernels, the most robust reusable physics API may be analytic derivatives such as:

```text
dpsi/dR_source
dpsi/dZ_source

dBr/dR_source
dBr/dZ_source

dBz/dR_source
dBz/dZ_source
```

These are useful even for users who never install JAX.

Where analytic derivatives are available and verifiable, they should remain first-class scientific formulas rather than being hidden behind autodiff.

The JAX implementation can then be tested against them.

---

## Initial benchmark targets

For the Green-function PoC, benchmark at least:

### Scalar/small-array evaluation

Confirm that JAX overhead makes these cases no worse than expected and document that ordinary NumPy should remain preferred.

### Large point-source response matrix

Representative shapes should include the sensor/source counts used by VFIT and VAFT electromagnetic workflows.

Measure:

- first-call compile time;
- steady-state CPU execution time;
- memory use where practical;
- NumPy/JAX numerical parity.

### Repeated moving-source evaluation

This is the most relevant nonlinear-inverse workload.

Benchmark repeated recomputation of the response matrix as a small number of source positions change.

### Batched states

Evaluate whether `vmap` across many parameter states or time slices provides a useful scaling advantage.

Do not interpret microbenchmarks alone as sufficient reason to broaden the backend.

---

## Acceptance criteria

- [ ] Default VAFT installation remains JAX-free.
- [ ] Default VAFT import does not import JAX.
- [ ] Python 3.12 JAX environment is documented and reproducible.
- [ ] A separate optional JAX CI job exists.
- [ ] x64 behavior is explicitly enforced or validated in JAX-only tests.
- [ ] At least one exact electromagnetic Green kernel has a JAX-compatible PoC.
- [ ] NumPy/JAX value parity is quantified over representative and edge-case domains.
- [ ] Automatic derivatives are checked independently.
- [ ] First-call compilation cost and steady-state cost are reported separately.
- [ ] No global backend switch is introduced.
- [ ] OMAS/IMAS, database, plotting, workflow, and external-code layers remain JAX-independent.
- [ ] VFIT FourFilament or an equivalent real downstream inverse problem is used to decide whether expansion is justified.
- [ ] The final implementation documents clearly which VAFT kernels are JAX-compatible and which are not.

---

## Non-goals

This issue does **not**:

- make JAX a required VAFT dependency;
- replace NumPy/SciPy;
- rewrite the whole `vaft.formula` or `vaft.process` namespace;
- provide a global array-backend abstraction;
- make OMAS/IMAS objects JAX-compatible;
- JIT database, plotting, workflow, or external-code operations;
- replace SciPy sparse solvers;
- add GPU/CUDA/ROCm CI in the first stage;
- duplicate large parts of VAFT physics solely for backend symmetry;
- guarantee that every numerical API will eventually support JAX.

The intended outcome is much narrower:

> establish a clean opt-in Python 3.12 JAX environment and validate a small set of differentiable VAFT physics kernels so downstream inverse solvers can use them without contaminating the normal VAFT stack.

**Comment by HongSik-Yun-Fusion** [comment]
Premise check against PyPI (2026-09-18): the "Python 3.12 only" JAX environment is no longer forced by JAX.

- `jax` 0.11.2 declares `requires_python >=3.12`.
- `jaxlib` 0.11.2 ships CPU wheels for cp312, cp313, cp314 **and cp315**.

The issue also describes normal VAFT as "Python 3.10–3.13". That predates #1008 and #1009, which make 3.14 canonical and target 3.10–3.15. Suggested adjustments:

- Run the optional JAX CI job on the canonical **3.14** interpreter instead of pinning it to 3.12. That way it shares the NumPy/SciPy stack #1012 settles on, and NumPy/JAX parity is measured against the same reference the rest of CI uses.
- State the extra's contract as **Python ≥ 3.12**, because JAX excludes 3.10 and 3.11. That keeps the issue's rule against permissive markers that would silently install nothing on 3.10/3.11.
- Keep the `jax>=0.11,<0.12` range, reviewed at implementation time as the issue already says.

Nothing else in the design (no global backend, x64 enforced in JAX-only tests, Green-function PoC first) is affected.

**Comment by HongSik-Yun-Fusion** [comment]
Cross-link from the acceleration audit in #1013.

The current code structure supports keeping this issue narrowly focused on **differentiability**, rather than broadening JAX into VAFT's generic performance/GPU backend.

- `compute_point_response_matrices()` is already a fully broadcast NumPy/SciPy implementation after #219, so it is a good canonical parity baseline for the Green-function PoC.
- The most relevant JAX benchmark remains the one already described here: repeated moving-source response evaluation / batched states for VFIT, including first-call compile cost, steady-state `jit`, `vmap`, and Jacobian evaluation.
- General CPU hot-loop acceleration belongs in #1013 (Numba only when profiling justifies it).
- Generic GPU field-line / trajectory acceleration should be a separate follow-up only if a representative workload demonstrates a crossover benefit. It should not be used as a reason to add a global JAX/backend switch here.

So the intended separation remains:

```text
NumPy/SciPy : canonical production/reference path
Numba       : optional CPU JIT for measured hot loops (#1013)
JAX         : optional differentiable kernels / autodiff (#1002)
GPU         : workload-specific follow-up only after benchmark evidence
```

This also means a future GPU study does not need to change the architecture of this issue: JAX may execute on accelerators when useful, but accelerator support is not the scientific reason for the JAX layer.
